R/read_csv.R

```diff
@@ -52,253 +75,266 @@
-                                                                      along="chain")
-          }
-        }
-        if (length(variables) > 0) {
-          new_post_warmup_draws <- posterior::as_draws_array(draws[(num_warmup_draws+1):all_draws, variables])
-          if (is.null(post_warmup_draws)) {
-            post_warmup_draws <- new_post_warmup_draws
-          } else {
-            post_warmup_draws <- posterior::bind_draws(post_warmup_draws, new_post_warmup_draws, along="chain")
+          if (length(sampler_diagnostics) > 0) {
+            warmup_sampler_diagnostics_draws <- posterior::bind_draws(
+              warmup_sampler_diagnostics_draws,
+              posterior::as_draws_array(draws[1:num_warmup_draws, sampler_diagnostics]),
+              along="chain"
+            )
+            post_warmup_sampler_diagnostics_draws <- posterior::bind_draws(
+              post_warmup_sampler_diagnostics_draws,
+              posterior::as_draws_array(draws[(num_warmup_draws+1):all_draws, sampler_diagnostics]),
+              along="chain"
+            )
           }
+        } else {
+            warmup_draws <- NULL
+            warmup_sampler_diagnostics_draws <- NULL
+            if (length(variables) > 0) {
+              post_warmup_draws <- posterior::bind_draws(
+                post_warmup_draws,
+                posterior::as_draws_array(draws[, variables]),
+                along="chain"
+              )
+            }
+            if (length(sampler_diagnostics) > 0 && metadata$algorithm != "fixed_param") {
+              post_warmup_sampler_diagnostics_draws <- posterior::bind_draws(
+                post_warmup_sampler_diagnostics_draws,
+                posterior::as_draws_array(draws[, sampler_diagnostics]),
+                along="chain"
+              )
+            }
         }
-        if (length(sampler_diagnostics) > 0) {
-          new_post_warmup_sampler_diagnostics_draws <- posterior::as_draws_array(draws[(num_warmup_draws+1):all_draws, sampler_diagnostics])
-          if (is.null(post_warmup_sampler_diagnostics_draws)) {
-            post_warmup_sampler_diagnostics_draws <- new_post_warmup_sampler_diagnostics_draws
-          } else {
-            post_warmup_sampler_diagnostics_draws <- posterior::bind_draws(post_warmup_sampler_diagnostics_draws,
-                                                                          new_post_warmup_sampler_diagnostics_draws,
-                                                                          along="chain")
-          }
+      } else if (metadata$method == "variational") {
+        # ignore first line as it's just the mean and lp__ as it's always 0
+        variational_draws <- posterior::as_draws_matrix(
+          draws[-1, colnames(draws) != "lp__", drop=FALSE]
+        )
+        if ("log_p__" %in% posterior::variables(variational_draws)) {
+          variational_draws <- posterior::rename_variables(variational_draws, lp__ = "log_p__")
```

Comment:
I made renaming `log_p__` and `log_g__` conditional on them being included. If the user specifies the `variables` argument they might not be and then `posterior::rename_variables()` would error.

Reply:
Great catch!